treaded-subroutined.c

```diff
@@ -0,0 +1,478 @@
+        printf("PC out of bounds\n");
+        pcpu->state = Cpu_Break;
+        return Instr_Break;
+    }
+    return fetch(pcpu);
+}
+
+static inline decode_t decode(Instr_t raw_instr, const cpu_t *pcpu) {
+    assert(pcpu);
+    decode_t result = {0};
+    result.opcode = raw_instr;
+    switch (raw_instr) {
+    case Instr_Nop:
+    case Instr_Halt:
+    case Instr_Print:
+    case Instr_Swap:
+    case Instr_Dup:
+    case Instr_Inc:
+    case Instr_Add:
+    case Instr_Sub:
+    case Instr_Mul:
+    case Instr_Rand:
+    case Instr_Dec:
+    case Instr_Drop:
+    case Instr_Over:
+    case Instr_Mod:
+    case Instr_And:
+    case Instr_Or:
+    case Instr_Xor:
+    case Instr_SHL:
+    case Instr_SHR:
+    case Instr_Rot:
+    case Instr_SQRT:
+    case Instr_Pick:
+        result.length = 1;
+        break;
+    case Instr_Push:
+    case Instr_JNE:
+    case Instr_JE:
+    case Instr_Jump:
+        result.length = 2;
+        if (!(pcpu->pc+1 < PROGRAM_SIZE)) {
+            printf("PC+1 out of bounds\n");
+            result.length = 1;
+            result.opcode = Instr_Break;
+            break;
+        }
+        result.immediate = (int32_t)pcpu->pmem[pcpu->pc+1];
+        break;
+    case Instr_Break:
+    default: /* Undefined instructions equal to Break */
+        result.length = 1;
+        result.opcode = Instr_Break;
+        break;
+    }
+    return result;
+}
+
+static inline decode_t fetch_decode(cpu_t *pcpu) {
+    return decode(fetch_checked(pcpu), pcpu);
+}
+
+/*** Service routines ***/
+#define BAIL_ON_ERROR() if (cpu.state != Cpu_Running) break;
+
+typedef void (*service_routine_t)(cpu_t *pcpu, decode_t* pdecode);
+
+#define DISPATCH() do {\
+    goto *service_routines[decoded.opcode];   \
+   } while(0);
+
+/*
+#define DISPATCH() \
+__asm__ __volatile__("mov    (%0, %1, 8), %%rcx\n" \
+                     "jmpq   *%%rcx\n" \
+                     :: "r"(&service_routines), "r"((uint64_t)decoded.opcode): "%rcx");\
+__asm__ __volatile__("nopl 0x55667788(%%rax)":::"memory");
+*/
+
+#define ADVANCE_PC() \
+    cpu.pc += decoded.length;\
+    cpu.steps++; \
+    if (cpu.state != Cpu_Running || cpu.steps >= steplimit) break;
+
+static inline void push(cpu_t *pcpu, uint32_t v) {
+    assert(pcpu);
+    if (pcpu->sp >= STACK_CAPACITY-1) {
+        printf("Stack overflow\n");
+        pcpu->state = Cpu_Break;
+        return;
+    }
+    pcpu->stack[++pcpu->sp] = v;
+}
+
+static inline uint32_t pop(cpu_t *pcpu) {
+    assert(pcpu);
+    if (pcpu->sp < 0) {
+        printf("Stack underflow\n");
+        pcpu->state = Cpu_Break;
+        return 0;
+    }
+    return pcpu->stack[pcpu->sp--];
+}
+
+static inline uint32_t pick(cpu_t *pcpu, int32_t pos) {
+    assert(pcpu);
+    if (pcpu->sp - 1 < pos) {
+        printf("Out of bound picking\n");
+        pcpu->state = Cpu_Break;
+        return 0;
+    }
+    return pcpu->stack[pcpu->sp - pos];
+}
+
+cpu_t pcpu;
+decode_t pdecoded;
+
+sr_Nop:
```

Comment:
This label (and the rest below) is outside any scope. It won't compile as it is a syntactic error.